README.md
# hs-cmdstan


ChangeLog.md
# Changelog for hs-cmdstan

## Unreleased changes


test/Spec.hs
import CmdStan.SummaryParser
import CmdStan.Types
import Test.Hspec
import Control.Exception
import Text.Megaparsec
import Text.Megaparsec.Char
import Data.String.Here.Uninterpolated
import qualified Data.Map.Strict as Map

testParser :: Parsec String String a -> String -> IO a
testParser parser input =
  either (throwIO . userError . show) pure $ parse parser "" input

main :: IO ()
main = hspec $ do
  describe "parseInputFiles" $ do
    it "works for a single file" $
      testParser parseInputFiles "Input file: output.csv" `shouldReturn` ["output.csv"]
    it "works for a two files" $
      testParser parseInputFiles "Input files: output1.csv, output2.csv" `shouldReturn`
        ["output1.csv", "output2.csv"]

  describe "parseInferenceModel" $ do
    it "works" $
      testParser parseInferenceModel "Inference for Stan model: bernoulli_model"
        `shouldReturn` "bernoulli_model"

  describe "parseCounts" $ do
    it "works with 1 count" $
        testParser parseCounts "(1000)" `shouldReturn` [1000]
    it "works with 2 counts" $
        testParser parseCounts "(1000, 2000)" `shouldReturn` [1000, 2000]

  describe "parseChainInfo" $ do
    it "works with 1 chain" $ do
      let input = "1 chains: each with iter=(1000); warmup=(0); thin=(1); 1000 iterations saved."

      testParser parseChainInfo input `shouldReturn` ChainInfo
        { chainCount      = 1
        , chainIterCounts = [1000]
        , warmupCounts    = [0]
        , thinCounts      = [1]
        , savedIterations = 1000
        }

    it "works with 2 chain" $ do
      let input = "2 chains: each with iter=(1000,2000); warmup=(0,1); thin=(1,2); 3000 iterations saved."

      testParser parseChainInfo input `shouldReturn` ChainInfo
        { chainCount      = 2
        , chainIterCounts = [1000, 2000]
        , warmupCounts    = [0, 1]
        , thinCounts      = [1, 2]
        , savedIterations = 3000
        }

  describe "parseTimes" $ do
    it "works with 1 count" $
        testParser parseTimes "0.0070" `shouldReturn` [0.0070]
    it "works with 2 counts" $
        testParser parseTimes "(0.0070, 0.0080)" `shouldReturn` [0.0070, 0.0080]

  describe "parseWarmupTimes" $ do
    it "works with 1 count" $
        testParser parseWarmupTimes "Warmup took 0.0070 seconds"
          `shouldReturn` [0.0070]
    it "works with 2 counts" $
        testParser parseWarmupTimes "Warmup took (0.0070, 0.0080) seconds, 0.014 seconds total"
          `shouldReturn` [0.0070, 0.0080]

  describe "parseSamplingTimes" $ do
    it "works with 1 count" $
        testParser parseSamplingTimes "Sampling took 0.017 seconds"
          `shouldReturn` [0.017]
    it "works with 2 counts" $
        testParser parseSamplingTimes "Sampling took (0.016, 0.017) seconds, 0.033 seconds total"
          `shouldReturn` [0.016, 0.017]

  describe "skipUntilLp" $
    it "works" $ do
      let input = [here|
Mean     MCSE  StdDev     5%   50%   95%    N_Eff  N_Eff/s    R_hat

lp__            -7.3  2.4e-02    0.71   -8.6  -7.0  -6.8      888    26907     1.00
      |]

      testParser (skipUntilLp >> string "lp__") input `shouldReturn` "lp__"

  describe "parseStanStatistic" $
    it "works" $ do
      let input = "lp__            -7.3  2.4e-02    0.71   -8.6  -7.0  -6.8      888    26907     1.00"
      testParser parseStanStatistic input `shouldReturn`
        ("lp__", StanStatistic
          { mean         = -7.3
          , mcse         = 2.4e-02
          , stdDev       = 0.71
          , percents     = [-8.6, -7.0, -6.8]
          , numberOfEff  = 888
          , effPerSecond = 26907
          , rHat         = 1.00
          }
        )

  describe "parseStatisticMap" $ it "works" $ do
      let input = "lp__            -7.3  2.4e-02    0.71   -8.6  -7.0  -6.8      888    26907     1.00\ntheta           0.25  5.8e-03    0.12  0.077  0.23  0.45      401    23611      1.0\n"

      testParser parseStatisticMap input `shouldReturn` Map.fromList
        [ ("lp__", StanStatistic
            { mean         = -7.3
            , mcse         = 2.4e-02
            , stdDev       = 0.71
            , percents     = [-8.6, -7.0, -6.8]
            , numberOfEff  = 888
            , effPerSecond = 26907
            , rHat         = 1.00
            }
          )
        , ("theta", StanStatistic
            { mean         = 0.25
            , mcse         = 5.8e-03
            , stdDev       = 0.12
            , percents     = [0.077,  0.23,  0.45]
            , numberOfEff  = 401
            , effPerSecond = 23611
            , rHat         = 1.00
            }
          )
        ]

  describe "parseSampler" $ it "works" $ do
    testParser parseSampler "Samples were drawn using hmc with nuts." `shouldReturn` "hmc with nuts"

  describe "parseAutcorreltations" $ it "works" $ do
    let input = [here|
Displaying the autocorrelations for chain 1:

  Lag          lp__ accept_stat__    stepsize__   treedepth__  n_leapfrog__   divergent__      energy__         theta
    0           1.0           1.0           1.0           1.0           1.0           nan           1.0           1.0
    1           0.4           0.0           1.0           0.0           0.1           nan           0.4           0.5
    |]

    fmap show (testParser parseAutcorreltations input) `shouldReturn` show (AutoCorrelations
      { chainIndex         = 1
      , lag                = [0, 1]
      , aLogProbability    = [1.0, 0.4]
      , aAcceptStat        = [1.0, 0.0]
      , aStepSize          = [1.0, 1.0]
      , aTreeDepth         = [1.0, 0.0]
      , aNumberOfLeapFrogs = [1.0, 0.1]
      , aDivergent         = [0.0/0, 0.0/0]
      , aEnergy            = [1.0, 0.4]
      , aParams            = Map.fromList [("theta", [1.0, 0.5])]
      })


  describe "parseStatisticMap" $ it "works" $ do
      let input = [here|
Input files: output1.csv, output2.csv
Inference for Stan model: bernoulli_model
2 chains: each with iter=(1000,1000); warmup=(0,0); thin=(1,1); 2000 iterations saved.

Warmup took (0.0070, 0.0070) seconds, 0.014 seconds total
Sampling took (0.016, 0.017) seconds, 0.033 seconds total

                Mean     MCSE  StdDev     5%   50%   95%    N_Eff  N_Eff/s    R_hat

lp__            -7.3  2.4e-02    0.71   -8.6  -7.0  -6.8      888    26907     1.00
accept_stat__   0.92  2.8e-03    0.12   0.65  0.97   1.0  1.9e+03  5.8e+04  1.0e+00
stepsize__      0.94  2.5e-03   0.080   0.87   1.0   1.0  1.0e+03  3.0e+04  2.6e+13
treedepth__      1.4  1.1e-02    0.48    1.0   1.0   2.0  1.8e+03  5.4e+04  1.0e+00
n_leapfrog__     2.5  3.0e-02     1.3    1.0   3.0   3.0  1.7e+03  5.2e+04  1.0e+00
divergent__     0.00      nan    0.00   0.00  0.00  0.00      nan      nan      nan
energy__         7.8  3.5e-02     1.0    6.8   7.5   9.8  8.6e+02  2.6e+04  1.0e+00

theta           0.25  4.6e-03    0.12  0.082  0.23  0.46      643    19477      1.0

Samples were drawn using hmc with nuts.
For each parameter, N_Eff is a crude measure of effective sample size,
and R_hat is the potential scale reduction factor on split chains (at
convergence, R_hat=1).
      |]

      show (parseStanSummary input) `shouldBe` show (Right StanSummary
          { sPercentiles     = [5,50,95]
          , inputFiles       = ["output1.csv", "output2.csv"]
          , inferenceModel   = "bernoulli_model"
          , chainInfo        = ChainInfo
            { chainCount      = 2
            , chainIterCounts = [1000, 1000]
            , warmupCounts    = [0, 0]
            , thinCounts      = [1, 1]
            , savedIterations = 2000
            }
          , warmupTimes      = [0.0070, 0.0070]
          , samplingTimes    = [0.016, 0.017]
          , logProbability   = StanStatistic
            { mean         = -7.3
            , mcse         = 2.4e-02
            , stdDev       = 0.71
            , percents     = [-8.6, -7.0, -6.8]
            , numberOfEff  = 888
            , effPerSecond = 26907
            , rHat         = 1.00
            }
          , acceptance       = StanStatistic
            { mean         = 0.92
            , mcse         = 2.8e-03
            , stdDev       = 0.12
            , percents     = [0.65,0.97,1.0]
            , numberOfEff  = 1900
            , effPerSecond = 58000.0
            , rHat         = 1.00
            }
          , stepSize         = StanStatistic
            { mean         =  0.94
            , mcse         = 2.5e-03
            , stdDev       = 0.080
            , percents     = [0.87,   1.0,   1.0]
            , numberOfEff  = 1.0e+03
            , effPerSecond = 3.0e+04
            , rHat         = 2.6e+13
            }
          , treeDepth        = StanStatistic
            { mean         = 1.4
            , mcse         = 1.1e-02
            , stdDev       = 0.48
            , percents     = [1.0, 1.0, 2.0]
            , numberOfEff  = 1.8e+03
            , effPerSecond = 5.4e+04
            , rHat         = 1.0e+00
            }
          , leapFrog         = StanStatistic
            { mean         = 2.5
            , mcse         = 3.0e-02
            , stdDev       = 1.3
            , percents     = [ 1.0, 3.0, 3.0]
            , numberOfEff  = 1700
            , effPerSecond = 5.2e+04
            , rHat         = 1.0e+00
            }
          , divergent        = StanStatistic
            { mean         = 0.00
            , mcse         = 0/0.0
            , stdDev       = 0.00
            , percents     = [0.00, 0.00, 0.00]
            , numberOfEff  = 0/0.0
            , effPerSecond = 0/0.0
            , rHat         = 0/0.0
            }
          , energy           = StanStatistic
            { mean         = 7.8
            , mcse         = 3.5e-02
            , stdDev       = 1.0
            , percents     = [6.8, 7.5, 9.8]
            , numberOfEff  = 8.6e+02
            , effPerSecond = 2.6e+04
            , rHat         = 1.0e+00
            }
          , paramStats       = Map.singleton "theta" $ StanStatistic
              { mean         = 0.25
              , mcse         = 4.6e-03
              , stdDev       = 0.12
              , percents     = [0.082, 0.23, 0.46]
              , numberOfEff  = 643
              , effPerSecond = 19477
              , rHat         = 1.00
              }

          , sampler          = "hmc with nuts"
          , autoCorrelations = Nothing
          , unparsed = input
          }
        :: Either String StanSummary)


src/CmdStan.hs
module CmdStan
  (
  -- * make
    makeDefaultMakeConfig
  , make
  -- * stanc
  , StancConfig (..)
  , makeDefaultStancConfig
  , stanc
  -- * Stan
  , Method (..)
  , Initialization (..)
  , StanExeConfig (..)
  , makeDefaultSample
  , stan
  -- * Summary
  , ChainInfo (..)
  , StanStatistic (..)
  , AutoCorrelations (..)
  , StanSummary (..)
  , StansummaryConfig (..)
  , makeDefaultSummaryConfig
  , stansummary
  -- * diagnose
  , diagnose
  ) where
import System.Process
import System.Exit
import Data.List
import System.FilePath.Posix
import Control.Exception
import System.IO
import Control.Monad
import CmdStan.SummaryParser
import CmdStan.Types
import System.Directory
import System.Environment

throwWhenExitFailure :: ExitCode -> IO ()
throwWhenExitFailure = \case
  ExitSuccess -> pure ()
  e@ExitFailure {} -> throwIO e

makeDefaultMakeConfig :: FilePath -> IO MakeConfig
makeDefaultMakeConfig rootPath = do
  cmdStanDir <- maybe (throwIO $ userError "CMDSTAN_DIR not defined") pure =<<
    lookupEnv "CMDSTAN_DIR"

  canonicalPath <- canonicalizePath rootPath

  pure $ MakeConfig
    { stancFlags = Nothing
    , userHeader = Nothing
    , jobCount   = Nothing
    , rootPath   = canonicalPath
    , cmdStanDir = cmdStanDir
    }

makeConfigToCmdLine :: MakeConfig -> String
makeConfigToCmdLine MakeConfig {..} = unwords
  [ maybe "" ("USER_HEADER="<>) userHeader
  , maybe "" (\x -> "STANCFLAGS=\"" <> stancConfigToCmdLine x <> "\"") stancFlags
  , "make"
  , maybe "" (\x -> "-j" <> show x) jobCount
  , rootPath
  ]

make :: MakeConfig -> IO ()
make config@MakeConfig {..} = throwWhenExitFailure <=< withCurrentDirectory cmdStanDir $
  system $ makeConfigToCmdLine config

makeDefaultStancConfig :: FilePath -> StancConfig
makeDefaultStancConfig modelFilePath = StancConfig
  { modelName       = Nothing
  , modelFilePath   = modelFilePath
  , outputCppFile   = dropExtension modelFilePath <.> "hpp"
  , allowUndefined  = False
  , includePaths    = []
  , useOpenCL       = False
  , autoFormat      = False
  , printCanocial   = False
  , printCpp        = False
  , allOptimization = False
  , warnUnitialized = False
  }

stancConfigToCmdLine :: StancConfig -> FilePath
stancConfigToCmdLine StancConfig {..} = unwords
  [ "stanc"
  , maybe "" ("--name=" <>) modelName
  , "--o=" <> outputCppFile
  , if allowUndefined then "--allow-undefined" else ""
  , "--include-paths=" <> intercalate "," includePaths
  , if useOpenCL then "--use-opencl" else ""
  , if autoFormat then "--auto-format" else ""
  , if printCanocial then "--print-canonical" else ""
  , if printCpp then "--print-cpp" else ""
  , if allOptimization then "--O" else ""
  , if warnUnitialized then "--warn-uninitialized" else ""
  , modelFilePath
  ]

stanc :: StancConfig -> IO ()
stanc = throwWhenExitFailure <=< system . stancConfigToCmdLine

makeDefaultSummaryConfig :: [FilePath] -> StansummaryConfig
makeDefaultSummaryConfig files = StansummaryConfig
  { sampleFiles = files
  , autocorr    = Nothing
  , csvFilePath = Nothing
  , percentiles = [5, 50, 95]
  , sigFigs     = Nothing
  }

stansummaryConfigToCmdLine :: StansummaryConfig -> [String]
stansummaryConfigToCmdLine StansummaryConfig {..} =
  filter (not . null)
  [ maybe "" (\x -> "--autocorr=" <> show x) autocorr
  , maybe "" ("--csv_filename=" <>) csvFilePath
  , if null percentiles then "" else "--percentiles=" <> intercalate "," (map show percentiles)
  , maybe "" (\x -> "--sig_figs=" <> show x) sigFigs
  ] ++ sampleFiles

stansummary :: StansummaryConfig -> IO StanSummary
stansummary config = do
  (exitCode, output, err) <- readProcessWithExitCode "stansummary" (stansummaryConfigToCmdLine config) ""
  hPutStrLn stderr err
  case exitCode of
    ExitFailure {} -> throwIO exitCode
    ExitSuccess {} -> either (throwIO . userError) pure $ parseStanSummary output

diagnose :: [FilePath] -> IO ()
diagnose sampleFilePaths = throwWhenExitFailure =<< system ("diagnose " <> unwords sampleFilePaths)

methodToCmdLine :: Method -> String
methodToCmdLine = \case
  Sample { .. } -> "sample" ++ maybe "" (" diagnostic_file=" <>) diagnosticFile
  Optimize -> "optimize"
  Variational { .. } ->  "variational" ++ maybe "" ("diagnostic_file=" <>) diagnosticFile
  GenerateQuantities -> "generate_quantities"
  Diagnose -> "diagnose"

initializationToCmdLine :: Initialization -> String
initializationToCmdLine x = "init=" <> case x of
  IRealValue value -> show value
  IZero -> "0"
  IFilePath filePath -> filePath

data StanExeConfig = StanExeConfig
  { method          :: Method
  , inputData       :: Maybe FilePath
  , output          :: Maybe FilePath
  , initialValues   :: Maybe Initialization
  , randomSeed      :: Maybe Int
  , refreshInterval :: Maybe Int
  , processId       :: Maybe Int
  }

makeDefaultSample :: FilePath -> Int -> StanExeConfig
makeDefaultSample rootPath chainIndex = StanExeConfig
  { method          = Sample Nothing
  , inputData       = Just $ rootPath <.> "json"
  , output          = Just $ rootPath <> "_" <> show chainIndex <.> "csv"
  , initialValues   = Nothing
  , randomSeed      = Nothing
  , refreshInterval = Nothing
  , processId       = Just chainIndex
  }

toStanExeCmdLine :: StanExeConfig -> String
toStanExeCmdLine StanExeConfig {..} = unwords
  [ methodToCmdLine method
  , maybe "" ("data file=" <>) inputData
  , maybe "" ("output file=" <>) output
  , maybe "" initializationToCmdLine initialValues
  , maybe "" (\x -> "random=" <> show x) randomSeed
  , maybe "" (\x -> "refresh=" <> show x) refreshInterval
  , maybe "" (\x -> "id=" <> show x) processId
  ]

stan :: FilePath -> StanExeConfig -> IO ()
stan exeFilePath config =
  throwWhenExitFailure =<< system (exeFilePath <> " " <> toStanExeCmdLine config)


src/CmdStan/Types.hs
module CmdStan.Types where
import GHC.Generics
import Data.Map.Strict (Map)

data MakeConfig = MakeConfig
  { stancFlags :: Maybe StancConfig
  , userHeader :: Maybe FilePath
  , jobCount   :: Maybe Int
  , cmdStanDir :: FilePath
  , rootPath   :: FilePath
  } deriving(Show, Eq, Ord, Generic)

data StancConfig = StancConfig
  { modelName       :: Maybe FilePath
  , modelFilePath   :: FilePath
  , outputCppFile   :: FilePath
  , allowUndefined  :: Bool
  , includePaths    :: [FilePath]
  , useOpenCL       :: Bool
  , autoFormat      :: Bool
  , printCanocial   :: Bool
  , printCpp        :: Bool
  , allOptimization :: Bool
  , warnUnitialized :: Bool
  } deriving(Show, Eq, Ord, Generic)

data ChainInfo = ChainInfo
  { chainCount      :: Int
  , chainIterCounts :: [Int]
  , warmupCounts    :: [Int]
  , thinCounts      :: [Int]
  , savedIterations :: Int
  } deriving (Show, Eq, Ord, Generic)

data StanStatistic = StanStatistic
  { mean         :: Double
  , mcse         :: Double
  , stdDev       :: Double
  , percents     :: [Double]
  , numberOfEff  :: Double
  , effPerSecond :: Double
  , rHat         :: Double
  } deriving (Show, Eq, Ord, Generic)

data AutoCorrelations = AutoCorrelations
  { chainIndex         :: Int
  , lag                :: [Double]
  , aLogProbability    :: [Double]
  , aAcceptStat        :: [Double]
  , aStepSize          :: [Double]
  , aTreeDepth         :: [Double]
  , aNumberOfLeapFrogs :: [Double]
  , aDivergent         :: [Double]
  , aEnergy            :: [Double]
  , aParams            :: Map String [Double]
  } deriving (Show, Eq, Ord)

data StanSummary = StanSummary
  { sPercentiles     :: [Int]
  , inputFiles       :: [FilePath]
  , inferenceModel   :: String
  , chainInfo        :: ChainInfo
  , warmupTimes      :: [Double]
  , samplingTimes    :: [Double]
  , logProbability   :: StanStatistic
  , acceptance       :: StanStatistic
  , stepSize         :: StanStatistic
  , treeDepth        :: StanStatistic
  , leapFrog         :: StanStatistic
  , divergent        :: StanStatistic
  , energy           :: StanStatistic
  , paramStats       :: Map String StanStatistic
  , sampler          :: String
  , autoCorrelations :: Maybe AutoCorrelations
  , unparsed         :: String
  } deriving (Show, Eq, Ord, Generic)

data StansummaryConfig = StansummaryConfig
  { sampleFiles :: [FilePath]
  , autocorr    :: Maybe Int
  , csvFilePath :: Maybe FilePath
  , percentiles :: [Int]
  , sigFigs     :: Maybe Int
  } deriving (Show, Eq, Ord, Generic)

data Method
  = Sample { diagnosticFile :: Maybe FilePath }
  | Optimize
  | Variational { diagnosticFile :: Maybe FilePath }
  | GenerateQuantities
  | Diagnose

data Initialization
  = IRealValue Double
  | IZero
  | IFilePath FilePath


src/CmdStan/SummaryParser.hs
module CmdStan.SummaryParser where
import CmdStan.Types
import Data.List
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Text.Megaparsec
import Text.Megaparsec.Char
import Control.Monad
import Control.Arrow
import qualified Text.Megaparsec.Char.Lexer as L
import Data.Function

myFloat :: Parsec String String Double
myFloat = L.signed space (try L.float <|> fmap fromIntegral (L.decimal :: Parsec String String Int) <|> string "nan" *> pure (0.0/0.0))

{-
Input file: output.csv
Inference for Stan model: bernoulli_model
1 chains: each with iter=(1000); warmup=(0); thin=(1); 1000 iterations saved.

Warmup took 0.0070 seconds
Sampling took 0.017 seconds

                Mean     MCSE   StdDev     5%   50%   95%    N_Eff  N_Eff/s    R_hat

lp__            -7.2  3.7e-02     0.69   -8.8  -7.0  -6.7      348    20472      1.0
accept_stat__   0.93  3.3e-03  1.1e-01   0.69  0.98   1.0  1.0e+03  6.1e+04  1.0e+00
stepsize__      0.95      nan  1.8e-15   0.95  0.95  0.95      nan      nan      nan
treedepth__      1.4  1.5e-02  4.9e-01    1.0   1.0   2.0  1.1e+03  6.2e+04  1.0e+00
n_leapfrog__     2.7  5.1e-02  1.6e+00    1.0   3.0   7.0  9.8e+02  5.8e+04  1.0e+00
divergent__     0.00      nan  0.0e+00   0.00  0.00  0.00      nan      nan      nan
energy__         7.7  4.8e-02  9.8e-01    6.8   7.4   9.7  4.2e+02  2.5e+04  1.0e+00

theta           0.25  5.8e-03     0.12  0.077  0.23  0.45      401    23611      1.0

Samples were drawn using hmc with nuts.
For each parameter, N_Eff is a crude measure of effective sample size,
and R_hat is the potential scale reduction factor on split chains (at
convergence, R_hat=1).

-}

percentilesParser :: Parsec String String [Int]
percentilesParser = do
  (try (string "Percentiles: "  *> L.decimal `sepBy1` char ',')) <|> pure [5, 50, 95]

-- Input file: output.csv
parseInputFiles :: Parsec String String [FilePath]
parseInputFiles = do
  void $ string "Input file: " <|> string "Input files: "
  someTill (satisfy $ const True)
    (lookAhead $ void (char ',') <|> void (char '\n') <|> eof) `sepBy1` (char ',' *> space)

parseInferenceModel :: Parsec String String String
parseInferenceModel = do
  void $ string "Inference for Stan model: "
  some $ anySingleBut '\n'

{-
1 chains: each with iter=(1000); warmup=(0); thin=(1); 1000 iterations saved.

Warmup took 0.0070 seconds
Sampling took 0.017 seconds

2 chains: each with iter=(1000,1000); warmup=(0,0); thin=(1,1); 2000 iterations saved.

-}
parseCounts :: Parsec String String [Int]
parseCounts = do
  void $ char '('
  result <- L.decimal `sepBy1` (char ',' >> space)
  void $ char ')'
  pure result

parseChainInfo :: Parsec String String ChainInfo
parseChainInfo = do
  chainCount <- L.decimal
  void $ string " chains: each with"
  space
  chainIterCounts <- string "iter=" *> parseCounts
  void $ char ';'
  space
  warmupCounts    <- string "warmup=" *> parseCounts
  void $ char ';'
  space
  thinCounts      <- string "thin=" *> parseCounts
  void $ char ';'
  void space
  savedIterations <- L.decimal
  void space
  void $ string "iterations saved."
  pure ChainInfo {..}

parseTimes :: Parsec String String [Double]
parseTimes = fmap pure myFloat <|> do
  void $ char '('
  result <- myFloat `sepBy1` (char ',' >> space)
  void $ char ')'
  pure result
{-
Warmup took 0.0070 seconds
Sampling took 0.017 seconds

Warmup took (0.0070, 0.0070) seconds, 0.014 seconds total
Sampling took (0.016, 0.017) seconds, 0.033 seconds total
-}
parseWarmupTimes :: Parsec String String [Double]
parseWarmupTimes = do
  -- TODO skip spaces
  void $ string "Warmup took "
  result <- fmap pure myFloat <|> parseTimes
  void $ some $ anySingleBut '\n'
  pure result

parseSamplingTimes :: Parsec String String [Double]
parseSamplingTimes = do
  void $ string "Sampling took "
  result <- fmap pure myFloat <|> parseTimes
  void $ some $ anySingleBut '\n'
  pure result

skipUntilLp :: Parsec String String ()
skipUntilLp = void $ skipSomeTill (satisfy $ const True) $ lookAhead $ string "lp__"

{-
                Mean     MCSE  StdDev     5%   50%   95%    N_Eff  N_Eff/s    R_hat

lp__            -7.3  2.4e-02    0.71   -8.6  -7.0  -6.8      888    26907     1.00
-}
parseStanStatistic :: Parsec String String (String, StanStatistic)
parseStanStatistic = do
  name <- someTill (satisfy $ const True) (lookAhead spaceChar) <?> "Statistic Name"
  space
  mean <- myFloat <?> "mean"
  space
  mcse <- myFloat <?> "MCSE"
  space
  stdDev <- myFloat <?> "stddev"
  space
  rHat:effPerSecond:numberOfEff:rest <- reverse <$> myFloat `sepBy1` hspace1 <?> "Other stats"
  let percents = reverse rest
  pure (name, StanStatistic{..})

parseStatisticMap :: Parsec String String (Map String StanStatistic)
parseStatisticMap = fmap Map.fromList $ fix $ \next -> do
  (fmap Just (try $ parseStanStatistic <* newline) <|> pure Nothing) >>= \case
    Just x -> fmap (x :) next
    Nothing -> pure []

parseSampler :: Parsec String String String
parseSampler = do
  void $ string "Samples were drawn using "
  result <- string "hmc with nuts"
  void $ char '.'
  pure result

{-
Displaying the autocorrelations for chain 1:

  Lag          lp__ accept_stat__    stepsize__   treedepth__  n_leapfrog__   divergent__      energy__         theta
    0           1.0           1.0           1.0           1.0           1.0           nan           1.0           1.0

-}

parseAutcorreltations :: Parsec String String AutoCorrelations
parseAutcorreltations = do
  chainIndex <- string "Displaying the autocorrelations for chain " *> L.decimal <* char ':'
  space
  "Lag":"lp__":"accept_stat__":"stepsize__":"treedepth__":"n_leapfrog__":"divergent__":"energy__":paramNames
    <- (someTill (satisfy $ const True) $ lookAhead spaceChar) `sepBy1` hspace1
  space
  lag:aLogProbability:aAcceptStat:aStepSize:aTreeDepth:aNumberOfLeapFrogs:aDivergent:aEnergy:paramValues <- transpose <$>
    (((hspace *> myFloat) `sepBy1` hspace1) `sepBy` newline)

  let aParams = Map.fromList $ zip paramNames paramValues

  pure AutoCorrelations {..}

{-
Input files: output1.csv, output2.csv
Inference for Stan model: bernoulli_model
2 chains: each with iter=(1000,1000); warmup=(0,0); thin=(1,1); 2000 iterations saved.

Warmup took (0.0070, 0.0070) seconds, 0.014 seconds total
Sampling took (0.016, 0.017) seconds, 0.033 seconds total

                Mean     MCSE  StdDev     5%   50%   95%    N_Eff  N_Eff/s    R_hat

lp__            -7.3  2.4e-02    0.71   -8.6  -7.0  -6.8      888    26907     1.00
accept_stat__   0.92  2.8e-03    0.12   0.65  0.97   1.0  1.9e+03  5.8e+04  1.0e+00
stepsize__      0.94  2.5e-03   0.080   0.87   1.0   1.0  1.0e+03  3.0e+04  2.6e+13
treedepth__      1.4  1.1e-02    0.48    1.0   1.0   2.0  1.8e+03  5.4e+04  1.0e+00
n_leapfrog__     2.5  3.0e-02     1.3    1.0   3.0   3.0  1.7e+03  5.2e+04  1.0e+00
divergent__     0.00      nan    0.00   0.00  0.00  0.00      nan      nan      nan
energy__         7.8  3.5e-02     1.0    6.8   7.5   9.8  8.6e+02  2.6e+04  1.0e+00

theta           0.25  4.6e-03    0.12  0.082  0.23  0.46      643    19477      1.0

Samples were drawn using hmc with nuts.
For each parameter, N_Eff is a crude measure of effective sample size,
and R_hat is the potential scale reduction factor on split chains (at
convergence, R_hat=1).

Displaying the autocorrelations for chain 1:

  Lag          lp__ accept_stat__    stepsize__   treedepth__  n_leapfrog__   divergent__      energy__         theta
    0           1.0           1.0           1.0           1.0           1.0           nan           1.0           1.0
    1           0.4          -0.0           1.0           0.0           0.1           nan           0.4           0.5
    2           0.2          -0.0           1.0          -0.0          -0.0           nan           0.2           0.2
    3           0.1          -0.0           1.0           0.0          -0.0           nan           0.1           0.1
    4           0.0          -0.0           1.0          -0.0          -0.0           nan           0.1           0.1
    5          -0.0          -0.0           1.0           0.0          -0.0           nan          -0.0           0.1
    6          -0.1          -0.0           1.0           0.0           0.0           nan          -0.1           0.1
    7          -0.1          -0.0           1.0          -0.1           0.0           nan          -0.1           0.1
    8          -0.0           0.0           1.0          -0.0           0.1           nan          -0.0           0.0
    9           0.0           0.1           1.0           0.0           0.0           nan          -0.0           0.0
   10          -0.0          -0.1           1.0          -0.0           0.0           nan          -0.0           0.0

-}

parseStanSummary :: String -> Either String StanSummary
parseStanSummary input = left show $ parse theParser "" input where
  theParser = do
    inputFiles     <- parseInputFiles <?> "Parse Input Files"
    space
    sPercentiles   <- percentilesParser <?> "Parse Percentiles"
    space
    inferenceModel <- parseInferenceModel <?> "Parse Inference Model"
    space
    chainInfo      <- parseChainInfo <?> "Parse Chain Info"
    space
    warmupTimes    <- parseWarmupTimes <?> "Parse Warmup Times"
    space
    samplingTimes  <- parseSamplingTimes <?> "Parse Sampling Times"
    skipUntilLp <?> "Skip until lp__"
    ("lp__"         , logProbability) <- parseStanStatistic <* newline <?> "Parse lp__"
    ("accept_stat__", acceptance    ) <- parseStanStatistic <* newline <?> "Parse accept_stat__"
    ("stepsize__"   , stepSize      ) <- parseStanStatistic <* newline <?> "Parse stepsize__"
    ("treedepth__"  , treeDepth     ) <- parseStanStatistic <* newline <?> "Parse treedepth__"
    ("n_leapfrog__" , leapFrog      ) <- parseStanStatistic <* newline <?> "Parse n_leapfrog__"
    ("divergent__"  , divergent     ) <- parseStanStatistic <* newline <?> "Parse divergent__"
    ("energy__"     , energy        ) <- parseStanStatistic <* newline <?> "Parse energy__"
    space
    paramStats     <- parseStatisticMap <?> "Parse parameter stats"
    space
    sampler        <- parseSampler <?> "Parse parameter stats"
    space
    autoCorrelations <- optional parseAutcorreltations <?> "Parse autocorrelation"
    let unparsed = input
    pure StanSummary {..}
